ocaml/xapi-guard/lib/server_interface.ml

```diff
@@ -129,7 +129,14 @@ let push_vtpm ~cache (vm_uuid, _timestamp, key) contents =
   let* () = with_xapi ~cache @@ VTPM.set_contents ~self ~contents in
   Lwt_result.return ()
 
-let serve_forever_lwt_callback_vtpm ~cache mutex persist vm_uuid _ req body =
+let read_vtpm ~cache (vm_uuid, _timestamp, key) =
```

Comment:
There is another get_contents above in push_vtpm, that might need some integration with this one, not sure

Reply:
As long as the push is not done concurrently (looking at the code I see an `iter_s`, so probably ok) this might be OK as is, the read-modify-write is an internal detail of how that push function works, and reads would always return the latest locally cached value for each separate state.

Although there might be *some* concurrent from having both the Watcher thread *and* the guest pushing data.
So I'd put a per-vtpm-uuid lock around push_vtpm around get_contents+set_contents to ensure nothing interrupts inbetween.i.e.:

* acquire_per_vtpm_uuid_lock
* get_contents
* set_contenst
* release_per_vtpm_uuid_lock

Previously there was no lock, because we relied on the client not doing more than 1 vTPM operation on the same vTPM at the time, but now internally xapi-guard might do this.

Reply:
> Previously there was no lock

There's a per-vm lock (line 145), but unfortunately the watcher thread escapes this

Reply:
Because reads always check for files on disk, writes only remove files once the contents are pushed into xapi, and writes are serialized, there's a guarantee that reads will always get the latest version while there are files on disk. And if there aren't the read will be successful or fail depending on the xapi availability, a before